Source/Python/WarpXWrappers.cpp

```diff
@@ -655,4 +656,99 @@ extern "C"
         auto parser_exe = parser.compileHost<1>();
         return parser_exe(warpx.gett_new(lev));
     }
+
+    void warpx_moveParticlesBetweenSpecies(const char* char_src_species_name,
+        const char* char_dst_species_name, const int lev)
+    {
+        auto & mypc = WarpX::GetInstance().GetPartContainer();
+        const std::string src_species_name(char_src_species_name);
+        auto & src_spc = mypc.GetParticleContainerFromName(src_species_name);
+        const std::string dst_species_name(char_dst_species_name);
+        auto & dst_spc = mypc.GetParticleContainerFromName(dst_species_name);
+
+        for (WarpXParIter pti(src_spc, lev); pti.isValid(); ++pti) {
+            auto& src_tile = src_spc.ParticlesAt(lev, pti);
+            auto& dst_tile = dst_spc.ParticlesAt(lev, pti);
+
+            auto src_np = src_tile.numParticles();
+            auto dst_np = dst_tile.numParticles();
+
+            dst_tile.resize(dst_np + src_np);
+            amrex::copyParticles(dst_tile, src_tile, 0, dst_np, src_np);
+        }
+
+        // clear the source species
+        src_spc.clearParticles();
+    }
+
+    void warpx_calcSchottkyWeight(const char* char_species_name,
+        const amrex::ParticleReal pre_fac, const int lev)
+    {
+        // get the particle container for the species of interest
+        auto & mypc = WarpX::GetInstance().GetPartContainer();
+        const std::string species_name(char_species_name);
+        auto & myspc = mypc.GetParticleContainerFromName(species_name);
+        auto particle_comps = myspc.getParticleComps();
```

Comment:
This should be `const auto & particle_comps` - no reason to copy it locally, or allow it to be modified.

Possibly applies to `plo` and `dxi` just below, too

Reply:
Agreed. Thanks for pointing it out.